ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndex.scala

```diff
@@ -44,142 +51,97 @@ private[apiserver] final class LedgerConfigurationSubscriptionFromIndex(
-        case Some(result) =>
-          logger.info(
-            s"Initial ledger configuration lookup found configuration ${result._2} at ${result._1}. Looking for new ledger configurations from this offset."
-          )
-          configFound(result._1, result._2)
-        case None =>
-          logger.info(
-            s"Initial ledger configuration lookup did not find any configuration. Looking for new ledger configurations from the ledger beginning."
-          )
-          latestConfigurationState.set(None -> None)
-      }(servicesExecutionContext)
-      .map(_ => startStreamingUpdates())(servicesExecutionContext)
-      .failed
-      .foreach { exception =>
-        readyPromise.tryFailure(exception)
-        logger.error("Could not load the ledger configuration.", exception)
-      }(servicesExecutionContext)
-
-    private def startStreamingUpdates(): Unit = {
-      val killSwitch = RestartSource
-        .withBackoff(
-          RestartSettings(
-            minBackoff = 1.seconds,
-            maxBackoff = 30.seconds,
-            randomFactor = 0.1,
-          )
-        ) { () =>
-          indexService
-            .configurationEntries(latestConfigurationState.get._1)
-            .map {
-              case (offset, domain.ConfigurationEntry.Accepted(_, config)) =>
-                logger.info(s"New ledger configuration $config found at $offset")
-                configFound(offset, config)
-              case (offset, domain.ConfigurationEntry.Rejected(_, _, _)) =>
-                logger.info(s"New ledger configuration rejection found at $offset")
-                latestConfigurationState.updateAndGet(previous => Some(offset) -> previous._2)
-                ()
-            }
-        }
-        .viaMat(KillSwitches.single)(Keep.right[NotUsed, UniqueKillSwitch])
-        .toMat(Sink.ignore)(Keep.left[UniqueKillSwitch, Future[Done]])
-        .run()(materializer)
-      if (
-        !state.compareAndSet(SubscriptionState.LookingUp, SubscriptionState.Streaming(killSwitch))
-      ) {
-        killSwitch.shutdown()
+    private val killSwitch = RestartSource
+      .withBackoff(
+        RestartSettings(
+          minBackoff = 1.seconds,
+          maxBackoff = 30.seconds,
+          randomFactor = 0.1,
+        )
+      ) { () =>
+        indexService
+          .configurationEntries(latestConfigurationState.get._1)
+          .map {
+            case (offset, domain.ConfigurationEntry.Accepted(_, config)) =>
+              logger.info(s"New ledger configuration $config found at $offset")
+              scheduledTimeout.cancel()
+              latestConfigurationState.set(Some(offset) -> Some(config))
+              readyPromise.trySuccess(())
+              ()
+            case (offset, domain.ConfigurationEntry.Rejected(_, _, _)) =>
+              logger.info(s"New ledger configuration rejection found at $offset")
+              latestConfigurationState.updateAndGet(previous => Some(offset) -> previous._2)
```

Comment:
any reason for the `updateAndGet` instead of `set` as used previosly?

Reply:
Yes; if we get a rejection but we previously had a configuration, we need to keep the previous configuration instead of setting it back to `None`.

(This code was just moved.)

Reply:
I think it's because `set` doesn't allow you to access the previous value. I'd propose to use a proper case class for the state though.

Reply:
I'll do it in a new changeset.

Reply:
#10508 introduces a `State` enumeration.